# Autoscout24, Audi A4

Besteht eine Abhängigkeit zwischen dem Antrieb (Allrad, Vorderradantrieb) und dem Preis.

## Libraries and settings

In [17]:
# Libraries
import os
import pandas as pd
import matplotlib.pyplot as plt

# Ignore warnings
import warnings
warnings.filterwarnings("ignore")

# Show current working directory
print(os.getcwd())

/workspaces/project_arm_autoscout_audia1


## Import data

In [3]:
## Read data
df = pd.read_csv("./data/as24_a4.csv", sep=",")
df.head()

web-scraper-order                              web-scraper-start-url  \
0      1734006862-1  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   
1      1734006865-2  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   
2      1734006867-3  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   
3      1734006869-4  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   
4      1734006872-5  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   

   as24_a1                                       as24_a1-href  \
0      NaN  https://www.autoscout24.ch/de/d/audi-a4-avant-...   
1      NaN  https://www.autoscout24.ch/de/d/audi-a4-avant-...   
2      NaN  https://www.autoscout24.ch/de/d/audi-a4-cabrio...   
3      NaN  https://www.autoscout24.ch/de/d/audi-a4-avant-...   
4      NaN  https://www.autoscout24.ch/de/d/audi-a4-avant-...   

                                      title_raw  \
0         AUDI A4 Avant 40 TDI quattro S-tronic   
1         AUDI A4 Avant 1.4 TFSI Sport S-Tronic   
2                   AUDI A4 Cabriolet 1.8 Turbo   
3  AUDI A4 Avant 2.0 TDI Sport quattro S-tronic   
4   AUDI A4 Avant 40 TDI Sport quattro S-tronic   

                               short_description_raw     price_raw  \
0  // Kundenauftrag 079 887 62 06 // CH-Fahrzeug ...  CHF 21'999.–   
1                     Im Kundenauftrag zu verkaufen!  CHF 21'990.–   
2                                                NaN   CHF 7'900.–   
3                                                NaN  CHF 13'900.–   
4                           Sehr gepflegtes Fahrzeug  CHF 27'800.–   

                                        car_data_raw first_release_raw  \
0  Calendar icon10.2019Car iconKombiRoad icon136'...           10.2019   
1  Calendar icon07.2018Car iconKombiRoad icon79'1...           07.2018   
2  Calendar icon08.2003Car iconCabrioletRoad icon...           08.2003   
3  Calendar icon12.2017Car iconKombiRoad icon218'...           12.2017   
4  Calendar icon08.2019Car iconKombiRoad icon107'...           08.2019   

  mileage_raw  ... fuel_consumption_raw motor_vehicle_inspection_raw  \
0  136'000 km  ...         6.1 l/100 km                           Ja   
1   79'116 km  ...         5.5 l/100 km                           Ja   
2  100'000 km  ...         8.7 l/100 km                           Ja   
3  218'000 km  ...         4.5 l/100 km                           Ja   
4  107'000 km  ...           5 l/100 km                           Ja   

  accidentcar_raw guarantee_raw  \
0            Nein          Nein   
1      01.07.2018          Nein   
2      28.11.2024          Nein   
3      23.08.2023          Nein   
4            Nein            Ja   

                                 car_description_raw  \
0                        Kundenauftrag 079 887 62 06   
1                     Im Kundenauftrag zu verkaufen!   
2                                                NaN   
3                     Top zustand Autobahn Kilometer   
4  Vielen Dank für Ihr Interesse an unserem Fahrz...   

                                       equipment_raw             seller_raw  \
0  AusstattungenOptionale AusstattungAdBlue Behäl...             uniAuto AG   
1  AusstattungenOptionale AusstattungAudi smartph...  Allenmoos Garage GmbH   
2  AusstattungenOptionale Ausstattung6-Stufen-Aut...           Bruno Keller   
3  AusstattungenOptionale AusstattungAlu-Räder 5-...               PTE GmbH   
4  AusstattungenOptionale AusstattungAblagenpaket...    Garage Gfeller GmbH   

                               selleraddress_raw sellerwebsite_raw  \
0                Moosburgstrasse 12, 9200 Gossau               NaN   
1              Hofwiesenstrasse 268, 8050 Zürich               NaN   
2  Ueberlandstr. 34, 8051 Zürich/ Schwamendingen               NaN   
3             Seetalstrasse 44, 6020 Emmenbrücke               NaN   
4                   Unterdorf 1, 8261 Hemishofen               NaN   

   seller_phonenumber_raw  
0   Verkauf 078 883 55 87  
1  Geschäft 044 311 62 63  
2     Mobil 079 441 71 24  
3  G

### 2. Data Preparation

2.1 Count and identify missing values and remove them: 
- Vorlesungunterlagen SWxxx, Slide XXX
- Exercises SW2, Task 2

2.2 Cound and identify duplicates and remove them:
- Vorlesungunterlagen SWxxx, Slide XXX
- Exercises SW2, Task 2

2.3 Create new variables
- Vorlesungunterlagen SWxxx, Slide XXX
- Exercises SW3, Task 1

2.4 Enrich the data with open data
- Vorlesungunterlagen SWxxx, Slide XXX
- Exercises SWxxx, Task xxx


2. Datenaufbereitung (z. B. Umgang mit fehlenden Werten, Duplikaten etc.):
Zu finden in Woche 02, Aufgabe 2:
- Übungen zur Prüfung und Bereinigung von Daten auf fehlende oder doppelte Werte.



Count numbers of rows and columns

In [4]:
# Dimension (rows, columns)
print('Dimension:', df.shape)

# Number of rows
print('Number of rows:', df.shape[0])

# Number of columns
print('Number of columns:', df.shape[1])

Dimension: (409, 25)
Number of rows: 409
Number of columns: 25


Get the datatype

In [5]:
# Get data types (note that in pandas, a string is referred to as 'object')
df.dtypes

web-scraper-order                object
web-scraper-start-url            object
as24_a1                         float64
as24_a1-href                     object
title_raw                        object
short_description_raw            object
price_raw                        object
car_data_raw                     object
first_release_raw                object
mileage_raw                      object
transmission_raw                 object
drive_raw                        object
cartype_raw                      object
fuel_raw                         object
ps_raw                           object
fuel_consumption_raw             object
motor_vehicle_inspection_raw     object
accidentcar_raw                  object
guarantee_raw                    object
car_description_raw              object
equipment_raw                    object
seller_raw                       object
selleraddress_raw                object
sellerwebsite_raw                object
seller_phonenumber_raw           object


In [6]:
# Count missing values
print(pd.isna(df).sum())

# Identify rows with missing values
df[df.isna().any(axis=1)].head()

web-scraper-order                 0
web-scraper-start-url             0
as24_a1                         409
as24_a1-href                      0
title_raw                         0
short_description_raw           103
price_raw                         0
car_data_raw                      0
first_release_raw                 0
mileage_raw                       0
transmission_raw                  0
drive_raw                         0
cartype_raw                       0
fuel_raw                          0
ps_raw                            0
fuel_consumption_raw              0
motor_vehicle_inspection_raw      0
accidentcar_raw                   0
guarantee_raw                     0
car_description_raw              97
equipment_raw                     8
seller_raw                        0
selleraddress_raw                 0
sellerwebsite_raw               268
seller_phonenumber_raw            0
dtype: int64


web-scraper-order                              web-scraper-start-url  \
0      1734006862-1  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   
1      1734006865-2  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   
2      1734006867-3  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   
3      1734006869-4  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   
4      1734006872-5  https://www.autoscout24.ch/de/s/mo-a4/mk-audi?...   

   as24_a1                                       as24_a1-href  \
0      NaN  https://www.autoscout24.ch/de/d/audi-a4-avant-...   
1      NaN  https://www.autoscout24.ch/de/d/audi-a4-avant-...   
2      NaN  https://www.autoscout24.ch/de/d/audi-a4-cabrio...   
3      NaN  https://www.autoscout24.ch/de/d/audi-a4-avant-...   
4      NaN  https://www.autoscout24.ch/de/d/audi-a4-avant-...   

                                      title_raw  \
0         AUDI A4 Avant 40 TDI quattro S-tronic   
1         AUDI A4 Avant 1.4 TFSI Sport S-Tronic   
2                   AUDI A4 Cabriolet 1.8 Turbo   
3  AUDI A4 Avant 2.0 TDI Sport quattro S-tronic   
4   AUDI A4 Avant 40 TDI Sport quattro S-tronic   

                               short_description_raw     price_raw  \
0  // Kundenauftrag 079 887 62 06 // CH-Fahrzeug ...  CHF 21'999.–   
1                     Im Kundenauftrag zu verkaufen!  CHF 21'990.–   
2                                                NaN   CHF 7'900.–   
3                                                NaN  CHF 13'900.–   
4                           Sehr gepflegtes Fahrzeug  CHF 27'800.–   

                                        car_data_raw first_release_raw  \
0  Calendar icon10.2019Car iconKombiRoad icon136'...           10.2019   
1  Calendar icon07.2018Car iconKombiRoad icon79'1...           07.2018   
2  Calendar icon08.2003Car iconCabrioletRoad icon...           08.2003   
3  Calendar icon12.2017Car iconKombiRoad icon218'...           12.2017   
4  Calendar icon08.2019Car iconKombiRoad icon107'...           08.2019   

  mileage_raw  ... fuel_consumption_raw motor_vehicle_inspection_raw  \
0  136'000 km  ...         6.1 l/100 km                           Ja   
1   79'116 km  ...         5.5 l/100 km                           Ja   
2  100'000 km  ...         8.7 l/100 km                           Ja   
3  218'000 km  ...         4.5 l/100 km                           Ja   
4  107'000 km  ...           5 l/100 km                           Ja   

  accidentcar_raw guarantee_raw  \
0            Nein          Nein   
1      01.07.2018          Nein   
2      28.11.2024          Nein   
3      23.08.2023          Nein   
4            Nein            Ja   

                                 car_description_raw  \
0                        Kundenauftrag 079 887 62 06   
1                     Im Kundenauftrag zu verkaufen!   
2                                                NaN   
3                     Top zustand Autobahn Kilometer   
4  Vielen Dank für Ihr Interesse an unserem Fahrz...   

                                       equipment_raw             seller_raw  \
0  AusstattungenOptionale AusstattungAdBlue Behäl...             uniAuto AG   
1  AusstattungenOptionale AusstattungAudi smartph...  Allenmoos Garage GmbH   
2  AusstattungenOptionale Ausstattung6-Stufen-Aut...           Bruno Keller   
3  AusstattungenOptionale AusstattungAlu-Räder 5-...               PTE GmbH   
4  AusstattungenOptionale AusstattungAblagenpaket...    Garage Gfeller GmbH   

                               selleraddress_raw sellerwebsite_raw  \
0                Moosburgstrasse 12, 9200 Gossau               NaN   
1              Hofwiesenstrasse 268, 8050 Zürich               NaN   
2  Ueberlandstr. 34, 8051 Zürich/ Schwamendingen               NaN   
3             Seetalstrasse 44, 6020 Emmenbrücke               NaN   
4                   Unterdorf 1, 8261 Hemishofen               NaN   

   seller_phonenumber_raw  
0   Verkauf 078 883 55 87  
1  Geschäft 044 311 62 63  
2     Mobil 079 441 71 24  
3  G

### Count and identify duplicate values

In [9]:
# # Count duplicated values
print(df.duplicated().sum())

# Identify rows with duplicated values, e.g.:
#df[df[['web-scraper-order', 'price_raw', 'address_raw']].duplicated()]

0


Save data to file

In [ ]:
df.to_csv('./Data/apartments_data_prepared.csv', 
          sep=",", 
          encoding='utf-8',
          index=False)

### 3. Data Storage

3.1 In a database like MySQL or PostgreSQL
- Vorlesungunterlagen SWxxx, Slide XXX
- Exercises SW 2, Task 4

3. Speicherung der Daten in Datenbanken wie SQLite, MySQL oder PostgreSQL:
Zu finden in Woche 02, Aufgabe 4:
- Speicherung und Abfrage von Daten in einer Datenbank mithilfe des Jupyter-Notebooks
`apartments_database_Python.ipynb`.


In [21]:
# Libraries
import os
import sqlite3
import fnmatch
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Ignore warnings
import warnings
warnings.filterwarnings("ignore")

# Function to close a sqlite db-connection
def check_conn(conn):
     try:
        conn.cursor()
        return True
     except Exception as ex:
        return False

# Get current working directory
print(os.getcwd())

/workspaces/project_arm_autoscout_audia1


Create sqlite data base

In [19]:
# Create data base
conn = sqlite3.connect('./Data/as24_a4.db') 

# Create cursor object to execute SQL commands
cursor = conn.cursor()

# Show dbs in the directory
flist = fnmatch.filter(os.listdir('.'), '*.db')
for i in flist:
    print(i)

OperationalError: unable to open database file

In [22]:
# Create a SQLite database connection
conn = sqlite3.connect('./Data/as24_a4.db')

# Store the dataframe in the SQLite database
df.to_sql('autoscout24_audi_a4', conn, if_exists='replace', index=False)

# Verify that the table has been created
query = "SELECT name FROM sqlite_master WHERE type='table';"
tables = pd.read_sql(query, conn)
print(tables)

# Close the connection
conn.close()

OperationalError: unable to open database file

Close db connection (if open)

In [ ]:
# Close db connection (if open)
try:
    if check_conn(conn):
        conn.close()
    else:
        pass
except:
    pass

# Status (True = open, False = closed)
print(check_conn(conn))

### 4. Exploratory data analysis (EDA)

4.1 Rich non-graphical and graphical exploratory data analysis (EDA)
- Vorlesungunterlagen SWxxx, Slide XXX
- Exercises SW3, Task 3
- Exercises SW 4, Task 3, 4


4. Explorative Datenanalyse (EDA):
Zu finden in Woche 04, Aufgabe 2:
- Identifikation und Entfernung von Ausreißern sowie Visualisierung von Statistiken und
Verteilungen.

Zu finden in Woche 05, Aufgabe 1:
- Nutzung von Python für grundlegende GIS-Funktionen und die Analyse räumlicher Daten.

Creating multiplots with .subplots()

In [ ]:
 # Define 2x2 subplots and whether x- and y-axes should be shared
fig, ax = plt.subplots(2, 2, figsize=(7,4), sharex=True, sharey=True)

ax[0, 0].scatter(x, y, s=100, marker='o', color='orangered') # row=0, col=0
ax[1, 0].scatter(x, z, s=100, marker='+', color='blue') # row=1, col=0
ax[0, 1].scatter(x, g, s=100, marker='*', color='deepskyblue') # row=0, col=1
ax[1, 1].scatter(x, e, s=100, marker='^', color='orange') # row=1, col=1

# Adding title
ax[0, 0].set_title('Figure at [0, 0]', fontsize = 12)
ax[1, 0].set_title('Figure at [1, 0]', fontsize = 12)
ax[0, 1].set_title('Figure at [0, 1]', fontsize = 12)
ax[1, 1].set_title('Figure at [1, 1]', fontsize = 12)

# Adding x-labels
ax[0, 0].set_xlabel('', fontsize = 10)
ax[1, 0].set_xlabel('x-label', fontsize = 10)
ax[0, 1].set_xlabel('', fontsize = 10)
ax[1, 1].set_xlabel('x-label', fontsize = 10)

# Adding y-labels
ax[0, 0].set_ylabel('n-label', fontsize = 10)
ax[1, 0].set_ylabel('y-label', fontsize = 10)
ax[0, 1].set_ylabel('', fontsize = 10)
ax[1, 1].set_ylabel('', fontsize = 10)

# Set the spacing between subplots
plt.subplots_adjust(left=0.1,
                    bottom=0.1, 
                    right=0.9, 
                    top=0.9, 
                    wspace=0.4, 
                    hspace=0.4)

plt.show()

### 5. Modeling Method

5.1 Use of multiple linear regression, regression tree or random forest regression as the modeling method
- Vorlesungunterlagen SWxxx, Slide XXX
- Exercises SW6, Task 2-5
- Exercises SW7, Task 1-2

5. Verwendung von Modellierungsmethoden (z. B. multiple lineare Regression, Entscheidungsbaum,
Random Forest):
Zu finden in Woche 07, Aufgaben 1 und 2:
- Durchführung von linearer Regression und Analyse von Entscheidungsbäumen sowie Random
Forest.

### 6. Model evaluation

6.1 Model evaluation using suitable measures of fit (e.g. r-squared, rmse)
- Vorlesungunterlagen SWxxx, Slide XXX
- - Exercises SW7, Task 1-2


6. Modellevaluation (z. B. R-Quadrat, RMSE):
Zu finden in Woche 07, Aufgabe 1 (b, c):
- Evaluation der Modelle mithilfe von R-Quadrat und Residuen.


### 7. Interpretation

7.1 Correct interpretation of model results and measures of fit
- Vorlesungunterlagen SWxxx, Slide XXX
- Exercises SW7, Task 2


7. Korrekte Interpretation von Modellergebnissen und Gütemaßen:
Zu finden in Woche 07, Aufgaben 1 und 2:
- Interpretation der Ergebnisse in den Bereichen lineare Regression und Random Forest.


8. Bereitstellung der Materialien in Jupyter-Notebooks:
In allen Wochen enthalten:
- Jede Aufgabe verlangt die Einreichung von Jupyter-Notebooks als `.html`-Dateien mit
entsprechenden Erklärungen.

1. Kreativität der Umsetzung:
- Hängt davon ab, wie einzigartig und innovativ du die Aufgaben kombinierst und erweiterst.
2. Verwendung von MySQL- oder PostgreSQL-Datenbanken:
Zu finden in Woche 02, Aufgabe 4:
- Standardmäßig wird SQLite verwendet, die Anpassung auf MySQL/PostgreSQL könnte jedoch
Kreativität zeigen.
3. Integration und Visualisierung geografischer Daten:
Zu finden in Woche 05, Aufgaben 1 und 3:
- GIS-Funktionen und Analyse nächster Nachbarn.
4. Verwendung des Chi-Quadrat-Tests oder ANOVA:
Zu finden in Woche 06, Aufgaben 3 und 4:
- Aufgaben mit Fokus auf einseitige Varianzanalyse (ANOVA) und Chi-Quadrat-Tests.
5. Verwendung von mehr als einem Modell für Vorhersagen:
Zu finden in Woche 07, Aufgaben 1 und 2:
- Einsatz von mehreren Modellen wie linearer Regression und Random Forest.
Projektpräsentation in Videoform (Maximal 3 Punkte):
Die Präsentation sollte den Inhalt und die Struktur der oben genannten Aufgaben klar und
überzeugend darstellen.